docs/tech-spike-write-ups/get-an-identity-oidc-oauth-with-ruby.md

```diff
@@ -0,0 +1,126 @@
+
+It’s expected that, for sufficiently advanced use cases, projects end up owning every part of the dependency chain. However, this simple spike required a significant investment digging into the library code, devising monkey patches, or just figuring out how things work in lieu of documentation. This is a strong signal that the tooling requires significant investment.
+
+There is a high risk that DFE-Digital will have to fork and maintain these gems. Alongside being a big time investment, making changes requires deep knowledge of OAuth and OIDC. Moreover, maintaining our own forks also means staying afloat of changes in the specification, as well as new security advisories, CVEs.
+Risk #3: Burden to build and ops overhead
+It was felt during the spike that there is no “shortcut” to building a quick version of a SSO provider, with good documentation, staging environments, bespoke integrations that might be slightly different for each client, a support team, oncall rota, and so forth.
+
+The closest prior art to a project that tackles the same scope in DFE Digital is DFE Signin. The “gut feeling” after the spike is that there is nothing inherent about our usecase that would make it “easier for us.”
+
+Reaching an MVP state requires a significant investment, with “not much to show” until it is fully done, as a half-baked SSO integration is one half away from being usable.
+Alternatives to building a Single Sign-on
+In light of the risks discussed above, some alternatives were thought of that would still solve the problems of code reuse / joined up user experience, while avoiding some of the thorny bits of OAuth and OIDC.
+
+# Approach #1: JWT token handover
+JSON Web Tokens (JWT) are a secure way of passing messages between two parties. A simple token exchange flow could be implemented both in Find a lost TRN and any client apps that would allow users to transfer data between the services in a secure way, such as their TRNs.
+
+## Quick pros:
+
+Simple to set up, does not require investment into learning the OAuth spec
+Working with JWT is a much more common skill in full stack devs, vs building an OIDC provider
+Can be packaged as a lightweight gem for clients
+
+## Quick cons:
+
+Will require sharing a private key between multiple projects; any leak would lead to having to perform a full rotation on all connected services
+But since we own all these services, that’s reasonable to do; OAuth was specifically designed to mitigate this issue in scenarios where 3rd parties are involved. We don’t have that need, so we don’t get the benefit from using OAuth (but still pay the price in additional complexity)
+Need to have a support process to identify and rotate leaked keys
+
+# Approach #2: Packaging up Find a lost TRN as a Rails engine
+Rather than passing the user around to a different service, we could package up the core parts of Find into a reusable module that can be plugged into other applications to give them similar functionality as needed.
```

Comment:
```suggestion
### Approach #2: Packaging up Find a lost TRN as a Rails engine
Rather than passing the user around to a different service, we could package up the core parts of Find into a reusable module that can be plugged into other applications to give them similar functionality as needed.
```